### ISSUE TITLE
Copy existing forum to a new domain

### ISSUE THREAD

**Issue by Lemmie17** [ilosuna/mylittleforum#552]
Hi all,

we would like to copy an existing mlf to a new domain for legal reasons. Backup exists, we would like to preserve internal links in the forum. Are there any how to papers or other checklists to pay attention to?

Best regards
Lemmie17

**Comment by loesler** [comment]
Hello,

in my opinion, you should transfer the full SQL dump of the forum and adjust the domain. Internal links must replaced manually.

/Micha

**Comment by auge8472** [comment]
The forum settings contains the absolute URL of the forum in the key `forum_address` and *possibly* an absolute URL of the home link in the key `forum_linkaddress` (could be a relative URL instead). Links inside the entries are text, so they can be relative links like `[msg=123]link to another entry[/msg]` or `[link=index.php?id=123]link to another entry[/link]` but also absolute URLs like `[link=https://example.com/forum/index.php?id=123]link to another entry[/link]` or pure text like `https://example.com/forum/index.php?id=123`.

If you want to ensure the forum internal links to be accessible under the new domain name you'll have to check all entries by hand or with a proper software. You can do this with an editor with search and replace functionallity. Open a copy of the backup of the entries table and search for forum URLs with your old domain name (i.e. `example.org/forum`) and replace the absolute URLs with relative ones including URL-parameters and possibly hashes (i.e. `index.php?id=123` or `index.php?id=123#123`).

**Comment by Lemmie17** [comment]
Hello loesler and auge872,

thanks for your hints! But before we can start doing this, we were confronted with another issue: the version of the MLF db to be migrated is 2.2.6 from 2011 - but the most recent MLF version is 2.4.24! There is a forum post about this matter: https://mylittleforum.net/forum/index.php?id=12236 - are there any other sources to be checked also before we go for this task?

Best regards,
Lemmie17

**Comment by loesler** [comment]
Hello,

I would update to the latest before transferring the forum. Some further information about updating can be found in the [wiki](https://github.com/ilosuna/mylittleforum/wiki/Update).

/Micha

**Comment by auge8472** [comment]
@loesler Do you talk about updating a forum **with** it's database content?

@Lemmie17 The forum changed it's database structure over the versions and a restoring a backup from 2.2.x into an installation of the current version (2.4.24) **will** fail. So you should backup the forum but only for safety reasons. Then you can update the forum step by step (1st step: update to 2.3.5, second step: update to 2.4.24)

**Comment by loesler** [comment]
Hello,

@auge8472 I would update the current installed version to the lastest one. Afterwards, I would try to transfer the DB via phpmyadmin (or a tool like this)

/Micha

**Comment by Lemmie17** [comment]
Hi auge872 and loesler,
thanks for your additional info -
our MLF is now migrated to version 2.4.24 successfully...
But there are some remaining questions - evtl. you have further hints?
- It was not possible to install latest PHP version 8 and also not latest supported PHP version 7.4 together with our MLF, because then the switch between different categories didn't worked. So we use now only version 7.3? Is this a known issue?
- In an old version of our forum on HTML basis we had the possibility, that every reply got an automatic "Re:" as prefix in the attention line. Could we install such a feature also with MLF?
- Is it possible to avoid the ® behind registered users?
Every further hints would be appreciated!
Best regards
Lemmie17

**Comment by auge8472** [comment]
Congratulation to the successfully migration from 2.2.6 to 2.4.24. :+1: 

- **PHP 7.4 and 8**: We made MLF compatible with PHP 7.***3*** by fixing incompatibilities in the MLF versions 2.4.21 and 2.4.22. I didn't made any tests with PHP 7.4 or 8 but I can't speak for @loesler at that point. So if the forum doesn't run under these PHP versions I can not confirm that fact but I also can not deny.
- **Re:**: The forum script does not contain such feature. In general it would be possible to add it but I would think it to have a high priority. It would be a simple task for a pull request. ;-)
- **®**: this is switchable as far as I know.

**Comment by loesler** [comment]
Hello,

at least one problem is [get_magic_quotes_gpc](https://www.php.net/manual/de/function.get-magic-quotes-gpc.php), which is depreciated. This function is called - if available - in [main.inc.php](https://github.com/ilosuna/mylittleforum/blob/master/includes/main.inc.php#L8). @auge8472 , whats happend, if we remove line 7 to 17?

In PHP8, I get the following error:
`Fatal error: Uncaught TypeError: error_reporting(): Argument #1 ($error_level) must be of type ?int, string given in /modules/smarty/sysplugins/smarty_internal_templatebase.php:201 Stack trace: #0 /modules/smarty/sysplugins/smarty_internal_templatebase.php(201): error_reporting('E_ALL & ~E_NOTI...') #1 /modules/smarty/sysplugins/smarty_internal_templatebase.php(134): Smarty_Internal_TemplateBase->_execute(Object(Smarty_Internal_Template), NULL, NULL, NULL, 1) #2 /index.php(226): Smarty_Internal_TemplateBase->display('default/main.tp...') #3 {main} thrown in /modules/smarty/sysplugins/smarty_internal_templatebase.php on line 201`

The ® can be disabled in the [style.(min.)css](https://github.com/ilosuna/mylittleforum/blob/master/themes/default/style.css#L315)

/Micha

**Comment by loesler** [comment]
Hello,

today, I start a quick test to solve the PHP 7.4/8 issues. First, I removed the call of the depreciated function `get_magic_quotes_gpc` to remove the warning, since the function returns always `false`. The PHP warning was the reason for the switching issue. 

Secondary, I update to the lastest smarty version (3.1.39). This solves the fatal errors. If there are no conflicts of interest, I will create a pull request, that solves both issues.

/Micha

**Comment by loesler** [comment]
FYI: https://github.com/smarty-php/smarty/issues/642

**Comment by auge8472** [comment]
Because of reading ain [the german Wikipedia article about PHP](https://de.wikipedia.org/wiki/PHP) I found a line [in a code example](https://de.wikipedia.org/wiki/PHP#Kommentare) that states `#` not being a comment opening char anymore in PHP 8. So we have to face the fact, that alone this single issue makes _the forum software currently **not** being **compatible with PHP 8**_.

**Comment by loesler** [comment]
... a lot of work.

**Comment by Lemmie17** [comment]
Hello again,

many thanks for your thoughts.

We are now about to change to productive our migrated MLF as described above. But one issue is left - when I try to create new users, the resulting links in my mailbox are like https://forum.our-new-domain.deindex.php?mode=register&id=95&key=DF4kH4RisK9gMGXw7rnp - i.e. without a slash between 'de' and index. So the registration of new users doesn't work this way... When I add the '/' manually to the link the registration the registration works. What could be the reason for this error?

Best regards, Giovanni

**Comment by loesler** [comment]
Hi,

the links is generated within the [registration](https://github.com/ilosuna/mylittleforum/blob/master/includes/register.inc.php#L187), i.e.,

`str_replace("[activate_link]", $settings['forum_address']."index.php?mode=register`

As one can see, there is no slash defined between the `forum address` and the `index.php`. So, how do you specify the forum address in your setting? Do you use `https://foum.example.org` or `https://foum.example.org/`? If you used the first version without slash, what is happend, if you changed the URI by adding a slash?

/Micha

**Comment by auge8472** [comment]
> … if you changed the URI by adding a slash?

Text snippet from the forum settings:
~~~
forum_address_desc =              'URL of the forum (use this format: http://www.domain.tld/forum/)'
~~~
We name the trailing slash in the description of the setting but we do not explicitely tell that it is *mandatory*.

Can we ensure the use of the correct URL-format (with trailing slash)? Maybe it's possible with a regular expression in a pattern attribute.

see therefore: [WhatWG specification](https://html.spec.whatwg.org/multipage/forms.html#client-side-form-validation) (en), [SelfHTML Eingabemuster mit pattern](https://wiki.selfhtml.org/wiki/HTML/Tutorials/Formulare/browsereigene_Validierung#Eingabemuster_mit_pattern) (de), browser support [according to CanIUse?](https://caniuse.com/input-pattern) (en)

**Comment by loesler** [comment]
Hello,

> Maybe it's possible with a regular expression in a pattern attribute.

Of course, something like `^http.*/$` will works. Good point!

/Micha

**Comment by Lemmie17** [comment]
Hello loesler and auge8472, 

this was it... thanks again.

> As one can see, there is no slash defined between the `forum address` and the `index.php`. So, how do you specify the forum address in your setting? Do you use `https://foum.example.org` or `https://foum.example.org/`? If you used the first version without slash, what is happend, if you changed the URI by adding a slash?

When the PHP 7.4 issue is solved, we will activate it on our web server... Meanwhile this thread can be closed!

Best regards, Lemmie17

**Comment by auge8472** [comment]
> loesler added "bug" "dependency" "next major version" "question" labels yesterday

Wow! Tag flodding. :-))
